### ISSUE TITLE
building from source does not like if you `git init` the tree.

### ISSUE THREAD

**Issue by pnkfelix** [rust-lang/rust#111766]
I tried this shell interaction:

```
% curl -O https://static.rust-lang.org/dist/rustc-1.69.0-src.tar.gz
% tar xzf rustc-1.69.0-src.tar.gz
% cd rustc-1.69.0-src/
% git init
% ./x.py setup compiler
% ./x.py build
```

I expected to see this happen: A successful bootstrap.
Instead, this happened:  the last step there fails with:

```
Building bootstrap
    Finished dev [unoptimized] target(s) in 0.04s
thread 'main' panicked at 'command did not execute successfully: cd "/home/ubuntu/Dev/Rust/rustc-1.69.0-src" && "git" "config" "--file" "/home/ubuntu/Dev/Rust/rustc-1.69.0-src/.gitmodules" "--get-regexp" "path"
expected success, got: exit status: 1', lib.rs:631:22
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

This is because, I presume, our logic for inferring whether to automatically update submodules does so based on whether it sees a `.git` subdirectory (which `git init` above will create).

One can work around this by adding this to the `config.toml` generated by `x.py setup`:

```
[build]
submodules = false
```

but it would be nice to not have to workaround it at all...

(using `git init` as a cheap way to checkpoint bits of state locally before I edit them is a personal  development pattern of mine, and I wouldn't be surprised if others use it as well...)

### Meta

`rustc --version --verbose`:
```
<version>
```

<details><summary>Backtrace</summary>
<p>

```
<backtrace>
```

</p>
</details>

**Comment by jyn514** [comment]
I don't think it makes sense to special case this in bootstrap. It seems pretty reasonable to expect the git repo to be the history for the rust project and not a random empty history ...


Why did you want to use a source tarball instead of a git checkout?

**Comment by pnkfelix** [comment]
> Why did you want to use a source tarball instead of a git checkout?

Putting my "random developer hat on": because a source tarball is what is supplied by the Rust project on the www.rust-lang.org website.

We don't point people there to the github repo at the sole option for developers. We offer them tarballs, for whatever reason. And then companies may well archive those tarballs internally, so that they are not relying on external services to recover state.

----

Anyway, at this point I don't think we need to do much special casing here.

In particular, I *don't mind* that we error in this case.

But it might be nice if the error messages for each of the two problems (1. the submodule update 
failure, and 2. the llvm  CI download failure) each gave a hint that one can work around the problem with the corresponding setting in the config.toml.

**Comment by jyn514** [comment]
Right, I understand why we offer tarballs, it's the combination of the tarball *and* the empty git repo that I don't understand.

I think detecting this case on error will be harder than just changing the `is_git` logic to verify that `.gitmodules` exists, that would be my preferred fix if we're going to do this at all.

**Comment by pnkfelix** [comment]
> it's the combination of the tarball and the empty git repo that I don't understand.

as I said above:
> (using git init as a cheap way to checkpoint bits of state locally before I edit them is a personal development pattern of mine, and I wouldn't be surprised if others use it as well...)

it is not cheap to checkout our repo.

**Comment by jyn514** [comment]
You can use --depth 2 to avoid downloading the full history, which should be roughly as fast as downloading a tarball.
